pkg/controller/shared_state.go

```diff
@@ -1136,23 +1137,30 @@ func (c *SharedState) cleanUpAzVolumeAttachments(ctx context.Context, attachment
 	var err error
 
 	for _, attachment := range attachments {
-		var patchRequired bool
 		patched := attachment.DeepCopy()
 
-		// if caller is azdrivernode, don't append cleanup annotation
-		if (caller != azdrivernode && !metav1.HasAnnotation(patched.ObjectMeta, consts.CleanUpAnnotation)) ||
-			// replica attachments should always be detached regardless of the cleanup mode
-			((cleanUp == cleanUpAttachment || patched.Spec.RequestedRole == azdiskv1beta2.ReplicaRole) && !metav1.HasAnnotation(patched.ObjectMeta, consts.VolumeDetachRequestAnnotation)) {
-			patchRequired = true
-			if caller != azdrivernode {
-				patched.Status.Annotations = azureutils.AddToMap(patched.Status.Annotations, consts.CleanUpAnnotation, string(caller))
+		if attachment.Spec.RequestedRole == azdiskv1beta2.PrimaryRole {
+			if cleanUp == cleanUpAttachment && !volumeDetachRequested(patched) {
+				markDetachRequest(patched, caller)
+			} else if deleteRequested, _ := objectDeletionRequested(&attachment); !deleteRequested {
+				// if primary azvolumeattachments are being cleaned up for driver uninstall, issue a DELETE call and continue
+				// note that this DELETE request will remove AzVolumeAttachment CRI without detaching the volume from node
+				if err = c.cachedClient.Delete(ctx, patched); err != nil {
+					return err
+				}
+			}
+		} else {
+			// replica mount should always be detached in cleanup regardless to the cleanup mode
+			if !volumeDetachRequested(patched) {
+				markDetachRequest(patched, caller)
 			}
-			if cleanUp == cleanUpAttachment || patched.Spec.RequestedRole == azdiskv1beta2.ReplicaRole {
-				patched.Status.Annotations = azureutils.AddToMap(patched.Status.Annotations, consts.VolumeDetachRequestAnnotation, string(caller))
+			// append cleanup annotation to prevent replica recreations except for when the clean up was triggered by triggered by node controller due to node failure.
```

Comment:
Duplicatated "triggered by" here. I guess you mean: "... was triggered by node controller ..."

Reply:
Oops